gstreamer/src/caps_serde.rs

```diff
@@ -0,0 +1,206 @@
+        let mut tup = serializer.serialize_tuple(2)?;
+        tup.serialize_element(self.0)?;
+        // `CapsFeature` is not available in `gstreamer-rs` yet
+        // Fake the type for now and use `None` as a value
+        tup.serialize_element::<Option<Structure>>(&None)?;
+        tup.end()
+    }
+}
+
+impl Serialize for CapsRef {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let iter = self.iter();
+        let size = iter.size_hint().0;
+        if size > 0 {
+            let mut seq = serializer.serialize_seq(Some(size))?;
+            for structure in iter {
+                seq.serialize_element(&CapsItemSe(structure))?;
+            }
+            seq.end()
+        } else {
+            let seq = serializer.serialize_seq(None)?;
+            seq.end()
+        }
+    }
+}
+
+impl Serialize for Caps {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        self.as_ref().serialize(serializer)
+    }
+}
+
+// `CapsFeature` is not available in `gstreamer-rs` yet
+struct CapsItemDe(Structure);
+impl From<CapsItemDe> for Structure {
+    fn from(caps_item: CapsItemDe) -> Structure {
+        caps_item.0
+    }
+}
+
+struct CapsItemVisitor;
+impl<'de> Visitor<'de> for CapsItemVisitor {
+    type Value = CapsItemDe;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a tuple `(Structure, Option<CapsFeature>)`")
+    }
+
+    fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<Self::Value, A::Error> {
+        let structure = seq
+            .next_element::<Structure>()?
+            .ok_or(de::Error::custom("Expected a `Structure` for `Caps` item"))?;
+        // `CapsFeature` is not available in `gstreamer-rs` yet
+        // Fake the type for now and expect `None` as a value
+        let feature_option = seq
+            .next_element::<Option<Structure>>()?
+            .ok_or(de::Error::custom(
+                "Expected an `Option<CapsFeature>` for `Caps` item",
+            ))?;
+        if feature_option.is_some() {
+            Err(de::Error::custom(
+                "Found a value for `CapsFeature`, expected `None` (not implemented yet)",
+            ))
+        } else {
+            Ok(CapsItemDe(structure))
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for CapsItemDe {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<CapsItemDe, D::Error> {
+        deserializer.deserialize_tuple(2, CapsItemVisitor)
+    }
+}
+
+struct CapsVisitor;
+impl<'de> Visitor<'de> for CapsVisitor {
+    type Value = Caps;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a sequence of `(Structure, Option<CapsFeature>)`")
+    }
+
+    fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<Self::Value, A::Error> {
+        let mut caps = Caps::new_empty();
```

Comment:
Does this work with `gst::Caps::new_empty()` vs `gst::Caps::new_any()`? Both have no structures but they're different

Reply:
Oh, I didn't noticed. I need to check the difference.

Reply:
You can only really check by `is_any()` and `is_empty()`, and then handle those cases separate from everything else somehow.

Reply:
Would that mean using a boolean (e.g. `is_any`) when there is no structures? We could use something like this (schematically):

```
caps: (
   is_any: bool,
   structures: [Structure],
)
```
but that would mean it could be inconsistent when `is_any` is `true` and `structures` is not empty.

Reply:
Proposal below to make it nicer https://github.com/sdroege/gstreamer-rs/pull/106#issuecomment-408617017 . Note that `Something` is just a placeholder, maybe you can find a better name :)